enterprise/backend/src/metabase_enterprise/advanced_permissions/query_processor/middleware/permissions.clj

```diff
@@ -0,0 +1,103 @@
+            [metabase.public-settings.premium-features :as premium-features]
+            [metabase.query-processor.middleware.permissions :as qp.perms]
+            [metabase.util.i18n :refer [tru]]))
+
+(def ^:private max-rows-in-limited-downloads 10000)
+
+(defn- is-download?
+  "Returns true if this query is being used to generate a CSV/JSON/XLSX export."
+  [query]
+  (some-> query :info :context name (str/includes? "download")))
+
+(defn- tables->download-perms-set
+  "Given a sequence of `tables-or-ids` referenced by a query, return a set of permissions required to download the
+  results of the query at the level specified by `download-level` (either :full or :limited)."
+  [database-or-id tables-or-ids download-level]
+  (query-perms/tables->permissions-path-set
+   database-or-id
+   tables-or-ids
+   {:table-perms-fn (fn [& path-components] (apply perms/feature-perms-path :download download-level path-components))
+    :native-perms-fn (fn [db-id] (perms/native-feature-perms-path :download download-level db-id))}))
+
+(defn- query->source-table-ids
+  [query]
+  ;; Remove the :native key (containing the transpiled MBQL) so that this helper function doesn't think the query is
+  ;; a native query. Actual native queries will still be detected appropriately.
+  (query-perms/query->source-table-ids (dissoc query :native)))
+
+(defn- download-perms-set
+  "Returns a set of permissions that are required to download a given query."
+  [{query-type :type, database :database, :as query} download-level]
+  (cond
+    (empty? query)         #{}
+    (= query-type :native) #{(perms/native-feature-perms-path :download download-level database)}
+    (= query-type :query)  (tables->download-perms-set database (query->source-table-ids query) download-level)))
+
+(defn- current-user-download-perms-level
+  "Returns the download permissions level which the current user has for the given query. If no user is bound, defaults
+  to full download permissions."
+  [query]
+  (cond
+    (perms/set-has-full-permissions-for-set? @api/*current-user-permissions-set* (download-perms-set query :full))
+    :full
+
+    (perms/set-has-full-permissions-for-set? @api/*current-user-permissions-set* (download-perms-set query :limited))
+    :limited
+
+    :else
+    :none))
+
+(defn apply-download-limit
+  "Pre-processing middleware to apply row limits to MBQL export queries if the user has `limited` download perms. This
+  does not apply to native queries, which are instead limited by the [[limit-download-result-rows]] post-processing
+  middleware."
+  [{query-type :type, {original-limit :limit} :query, :as query}]
+  (if (and (is-download? query)
+           (premium-features/has-feature? :advanced-permissions)
+           (= query-type :query)
+           (= (current-user-download-perms-level query) :limited))
+    (assoc-in query
+              [:query :limit]
+              (reduce min (filter some? [original-limit max-rows-in-limited-downloads])))
```

Comment:
ah yeah, I was referencing [this code](https://github.com/metabase/metabase/blob/087175f51ed082ca5db5a20a9cc8e6ebb59729ab/shared/src/metabase/mbql/util.cljc#L619-L621) which uses `reduce` for a similar case. I think `reduce` is a lot faster than `apply` if you have a large lazy seq as the input, since `apply` has to realize the entire seq up front. But that doesn't apply here so I'll change it.